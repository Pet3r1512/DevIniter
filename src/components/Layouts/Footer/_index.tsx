import { Copyright } from "lucide-react";

export default function Footer() {
  return (
    <footer className="dark:bg-black bg-gray-light h-36 flex flex-col gap-y-1.5 items-center justify-center cursor-default dark:text-gray text-black-main text-lg drop-shadow-[-50px_0_50px_rgba(0,0,0,0.25)]">
      <p>Released under MIT license</p>
      <p data-testid="author" className="flex items-center gap-x-0.5">
        Copyright <Copyright /> 2024-present by{" "}
        <span className="bg-gradient-to-br from-secondary via-primary dark:to-white to-secondary inline-block text-transparent bg-clip-text font-bold">
          DevIniter.
        </span>
      </p>
    </footer>
  );
}
