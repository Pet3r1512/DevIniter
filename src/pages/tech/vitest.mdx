import Code from "../../components/Documents/Code"
import HighlightedText from "../../components/Documents/HighlightedText"
import CodeBlock from "../../components/Documents/CodeBlock.tsx"

# Vitest

## Introduction

**Vitest** is a next-generation testing framework powered by **Vite**. It includes a smart and instant watch mode and is compatible with **Jest**. Moreover, **TypeScript**, **ESM**, and **JSX** are also supported.

**React Testing Library** is a simple and complete React DOM testing utilities that encourage good testing practices.

## Reason for Mandatory Tool

**DevIniter** believes that testing is one of the most important processes in the entire product development lifecycle. Testing ensures that everything works as expected, and changes should only occur when you decide. Therefore, <Code>Vitest</Code> and <Code>React Testing Library</Code> are included in our templates.

## Why Vitest?

Vitest offers several advantages that make it a compelling choice for developers:

1. **Speed**: Built on Vite, Vitest leverages its fast bundling and hot module replacement capabilities, resulting in quick test runs and instant feedback during development.

2. **Compatibility**: Vitest is designed to be compatible with Jest, allowing developers to easily migrate existing tests without significant changes to their codebase.

3. **TypeScript Support**: With first-class TypeScript support, Vitest enables developers to write type-safe tests, reducing runtime errors and improving code quality.

4. **Rich Features**: Vitest comes with a variety of built-in features such as mocking, snapshot testing, and code coverage, providing a comprehensive testing solution out of the box.

5. **Modern Syntax**: The framework supports modern JavaScript and TypeScript syntax, making it easier for developers to write clean and maintainable tests.

6. **Community and Ecosystem**: As part of the Vite ecosystem, Vitest benefits from a growing community and a wealth of plugins and integrations, enhancing its functionality and ease of use.

By choosing Vitest, developers can enjoy a powerful, efficient, and modern testing experience that aligns with contemporary development practices.

## Simple Test

In this example, we will write a very simple test in order to verify output from a function: 

<CodeBlock lang="ts" filename="sum.ts" code={
`export function sum(a: number, b: number) { 
    return a + b 
}`
} />

<CodeBlock lang="ts" filename="sum.test.ts" code={
`import { expect, it } from 'vitest'
import { sum } from './sum.ts'

it('should return 3 when adding 1 and 2', () => {
    expect(sum(1, 2)).toBe(3)
})`
} />

## Tests For Components

If you want to make sure that a specific component or element is displayed correctly in DOM, you can use below way by using **React Testing Libirary**. For intance, you have a <Code>Title</Code> component to display a simple heading:

<CodeBlock lang="tsx" filename="components/title.tsx" code={
`export default function Title() {
    return (
        <h1 
            // Define a data-testid attribute to make it easier to select in tests
            data-testid="title" 
        >
            Welcome To DevIniter
        </h1>
    )
}
`} />

Then, you can write a test to make sure it appear in DOM:

<CodeBlock lang="tsx" filename="components/title.test.tsx" code={
`import { expect, it } from 'vitest'
import { render, screen } from "@testing-library/react";
import Title from "./title"

it("should render Title component in DOM" , async () => {
    // Render Title component
    render(<Title />)

    // Wait to get the element by data-testid
    const title = await screen.getByTestId("title")

    // Expect it to be displayed in DOM
    // by using toBeInTheDocument
    // which is provided by @testing-library/jest-dom
    expect(title).toBeInTheDocument()
})
`} />

## Official Documents

For furthur information or instructions, please visit React Testing Library and Vitest's official websites:

1. Vitest: <HighlightedText content={{href: "https://vitest.dev", text: "vitest.dev"}} />
2. React Testing Library Github: <HighlightedText content={{href: "https://github.com/testing-library/react-testing-library#readme", text: "github.com/testing-library/react-testing-library"}} />