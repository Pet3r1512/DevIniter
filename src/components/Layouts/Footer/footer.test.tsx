import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import Footer from "./_index";

describe("Footer", () => {
  it("should match snapshot", () => {
    const footer = render(<Footer />);
    expect(footer).toMatchSnapshot();
  });

  it("should render correct content", () => {
    const footer = render(<Footer />);

    const license = footer.getByText("Released under MIT license");
    expect(license).toBeInTheDocument();

    const author = footer.getByTestId("author");
    expect(author).toBeInTheDocument();
  });
});
