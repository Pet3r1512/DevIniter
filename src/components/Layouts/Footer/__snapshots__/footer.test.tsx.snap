// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <footer
        class="dark:bg-black bg-gray-light h-36 flex flex-col gap-y-1.5 items-center justify-center cursor-default dark:text-gray text-black-main text-lg drop-shadow-[-50px_0_50px_rgba(0,0,0,0.25)]"
      >
        <p>
          Released under MIT license
        </p>
        <p
          class="flex items-center gap-x-0.5"
          data-testid="author"
        >
          Copyright 
          <svg
            class="lucide lucide-copyright"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <path
              d="M14.83 14.83a4 4 0 1 1 0-5.66"
            />
          </svg>
           2024-present by
           
          <span
            class="bg-gradient-to-br from-secondary via-primary dark:to-white to-secondary inline-block text-transparent bg-clip-text font-bold"
          >
            DevIniter.
          </span>
        </p>
      </footer>
    </div>
  </body>,
  "container": <div>
    <footer
      class="dark:bg-black bg-gray-light h-36 flex flex-col gap-y-1.5 items-center justify-center cursor-default dark:text-gray text-black-main text-lg drop-shadow-[-50px_0_50px_rgba(0,0,0,0.25)]"
    >
      <p>
        Released under MIT license
      </p>
      <p
        class="flex items-center gap-x-0.5"
        data-testid="author"
      >
        Copyright 
        <svg
          class="lucide lucide-copyright"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="M14.83 14.83a4 4 0 1 1 0-5.66"
          />
        </svg>
         2024-present by
         
        <span
          class="bg-gradient-to-br from-secondary via-primary dark:to-white to-secondary inline-block text-transparent bg-clip-text font-bold"
        >
          DevIniter.
        </span>
      </p>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
